### PR TITLE
Add other key from index bound

### PR DESCRIPTION
**Pull request by tjoneslo**
During an Index scan, now pass the end of the scan key from `IndexCursorUnidirectional`, through `FDBStoreData` and into `FDBStorageDescription`. The key is used to set an upper (or lower) bound on the range requested from the FDB storage layer. Hopefully more restricted than the rest of the index.

### CODE REVIEW

File: fdb-sql-layer-core/src/main/java/com/foundationdb/server/store/format/FDBStorageDescription.java [hunk 1]
```diff
@@ -257,6 +261,11 @@ public void indexIterator(FDBStore store, Session session, FDBStoreData storeDat
                               boolean key, boolean inclusive, boolean reverse,
                               FDBScanTransactionOptions transactionOptions) {
         KeySelector ksLeft, ksRight;
+
+        byte[] endKey = null;
+        if (storeData.endKey.getDepth() > 0)
```
Comment (nathanlws) [inline fdb-sql-layer-core/src/main/java/com/foundationdb/server/store/format/FDBStorageDescription.java]: A key that has been nudged has a zero depth but is *also* not empty. I don't recall if we ever do that though.

Perhaps at least an `else { assert endKey.getEncodedSize() == 0; }` ?

File: fdb-sql-layer-core/src/main/java/com/foundationdb/server/store/format/FDBStorageDescription.java [hunk 2]
```diff
@@ -264,22 +273,38 @@ public void indexIterator(FDBStore store, Session session, FDBStoreData storeDat
...
+                } else {
+                    ksLeft = KeySelector.firstGreaterThan(prefixBytes);
+                }
                 ksRight = KeySelector.firstGreaterThan(packKey(storeData));
             } 
             else {
                 ksLeft = KeySelector.firstGreaterOrEqual(packKey(storeData));
-                ksRight = KeySelector.firstGreaterOrEqual(ByteArrayUtil.strinc(prefixBytes));
+                if (endKey != null) {
+                    ksRight = KeySelector.firstGreaterOrEqual(ByteArrayUtil.strinc(endKey));
```
Comment (nathanlws) [inline fdb-sql-layer-core/src/main/java/com/foundationdb/server/store/format/FDBStorageDescription.java]: An inclusive bound probably shouldn't be `strinc`-ed, right? And if so, sounds like at least 1 new test is needed.

File: fdb-sql-layer-core/src/main/java/com/foundationdb/server/store/format/FDBStorageDescription.java [hunk 3]
```diff
@@ -264,22 +273,38 @@ public void indexIterator(FDBStore store, Session session, FDBStoreData storeDat
...
+                    ksRight = KeySelector.firstGreaterOrEqual(ByteArrayUtil.strinc(prefixBytes));
+                }
             }
         }
         else {
             if (reverse) {
-                ksLeft = KeySelector.firstGreaterThan(prefixBytes);
-                ksRight = KeySelector.firstGreaterOrEqual(packKey(storeData));
+                if (endKey != null) {
+                    ksLeft = KeySelector.firstGreaterOrEqual(endKey);
```
Comment (nathanlws) [inline fdb-sql-layer-core/src/main/java/com/foundationdb/server/store/format/FDBStorageDescription.java]: We don't want OrEqual on a non-inclusive left bound, right?

**Comment by nathanlws** [comment]
Couple inline comments but otherwise looks straight forward.

What was compensating for the end bound not being passed all the way down before. Is there code that can be removed now?

**Comment by tjoneslo** [comment]
Prior to this change the request for an index scan would always scan from the start key to the end (or beginning) of the index, regardless of the end point. The goal is to reduce the size of the requested data coming back from the K/V store. All of the checking the SQL Layer is doing on the returned data is still required.

File: fdb-sql-layer-core/src/main/java/com/foundationdb/server/store/format/FDBStorageDescription.java (hunk 3 above)
Comment (tjoneslo) [inline fdb-sql-layer-core/src/main/java/com/foundationdb/server/store/format/FDBStorageDescription.java]: This is correct. Take a query of the form `SELECT id FROM t1 where id >= 3 ORDER BY id DESC`. If `t1` contains the ids 2,3,4,5, the query has the end key set to `3`. So `.firstGreaterThan` is id = 4.

**Comment by tjoneslo** [comment]
Added the assert per merge comments. Ignoring the second comment. The third is correctly coded.

File: fdb-sql-layer-core/src/main/java/com/foundationdb/server/store/format/FDBStorageDescription.java (hunk 3 above)
Comment (tjoneslo) [inline fdb-sql-layer-core/src/main/java/com/foundationdb/server/store/format/FDBStorageDescription.java]: The inclusive/exclusive flag is set based upon the start key request, but the flag setting from the IndexScan optimizer setup doesn't make it down to this level. So we have to assume exclusive.

Comment (nathanlws) [inline fdb-sql-layer-core/src/main/java/com/foundationdb/server/store/format/FDBStorageDescription.java]: That's awfully delicate. At least a comment is warranted.

**Comment by nathanlws** [comment]
> Prior to this change the request for an index scan would always scan from the start key to the end (or beginning) of the index, regardless of the end point.

Right. So somewhere about the index cursor level is, presumably, double checking if the key falls within the range. Couldn't that be eliminated now?

Another merge is fine but we at least need a handle on it.

**Comment by tjoneslo** [comment]
In `IndexCursorUnidirectional#next()` are calls to `beforeStart()` and `pastEnd()` which do this checking. But in order to remove those, the `FDBIterationHelper->FDBStorageDescription` for `indexIterator` API needs to have the endKey Inclusive/exclusive flag passed through. 

The problem is removing those checks breaks the Persistit storage engine since it still does an index scan with an unbounded end key. 

I'll add the `indexIterator` API change to the card I'm currently working on. And add a comment to the `IndexCursorUnidirectional` that we can remove the affected methods once Persistit goes away.

**Comment by nathanlws** [comment]
Perfect. Thanks for the additional detail, exactly what I was looking for.

Approve

**Pull request merged**